src/annotator/anchoring/pdf.js

```diff
@@ -0,0 +1,430 @@
+    const result = [];
+    while (left.length > 0 || right.length > 0) {
+      if (right.length) {
+        result.push(right.shift());
+      }
+      if (left.length) {
+        result.push(left.pop());
+      }
+    }
+    return result;
+  }
+
+  return findPage(position.start).then(({ index }) => sortPages(index));
+}
+
+/**
+ * Anchor a set of selectors to a DOM Range.
+ *
+ * @param {HTMLElement} root
+ * @param {Array} selectors - Selector objects to anchor
+ * @param {Object} options - Options to pass to selector anchoring
+ * @return {Promise<Range>}
+ */
+function anchor(root, selectors, options = {}) {
+  const position = selectors.find(s => s.type === 'TextPositionSelector');
+  const quote = selectors.find(s => s.type === 'TextQuoteSelector');
+
+  let result = Promise.reject('unable to anchor');
+
+  const checkQuote = range => {
+    if (quote && quote.exact !== range.toString()) {
+      throw new Error('quote mismatch');
+    }
+    return range;
+  };
+
+  if (position) {
+    result = result.catch(() => {
+      return findPage(position.start).then(({ index, offset, textContent }) => {
+        const page = getPage(index);
+        const start = position.start - offset;
+        const end = position.end - offset;
+        const length = end - start;
+
+        checkQuote(textContent.substr(start, length));
+
+        const anchor = new TextPositionAnchor(root, start, end);
+        return anchorByPosition(page, anchor, options);
+      });
+    });
+  }
+
+  if (quote) {
+    result = result.catch(() => {
+      if (
+        position &&
+        quotePositionCache[quote.exact] &&
+        quotePositionCache[quote.exact][position.start]
+      ) {
+        const { page, anchor } = quotePositionCache[quote.exact][
+          position.start
+        ];
+        return anchorByPosition(page, anchor, options);
+      }
+
+      return prioritizePages(position).then(pageIndices => {
+        return findInPages(pageIndices, quote, position);
+      });
+    });
+  }
+
+  return result;
+}
+
+/**
+ * Convert a DOM Range object into a set of selectors.
+ *
+ * Converts a DOM `Range` object into a `[position, quote]` tuple of selectors
+ * which can be saved with an annotation and later passed to `anchor` to
+ * convert the selectors back to a `Range`.
+ *
+ * @param {HTMLElement} root - The root element
+ * @param {Range} range
+ * @param {Object} options -
+ *   Options passed to `TextQuoteAnchor` and `TextPositionAnchor`'s
+ *   `toSelector` methods.
+ * @return {Promise<[TextPositionSelector, TextQuoteSelector]>}
+ */
+function describe(root, range, options = {}) {
+  const normalizedRange = new xpathRange.BrowserRange(range).normalize();
+
+  const startTextLayer = getNodeTextLayer(normalizedRange.start);
+  const endTextLayer = getNodeTextLayer(normalizedRange.end);
+
+  if (startTextLayer !== endTextLayer) {
+    return Promise.reject(
+      new Error('selecting across page breaks is not supported')
```

Comment:
This does seem like an OK reason to reject a Promise: you're unable to provide a value for the Promise to resolve to 👍